### PR TITLE
Enable parallel fusion

### PR DESCRIPTION
**Pull request by chelini**
Rewrite `scf.for` to `scf.forall` right after fusion. This is legal since we tile and fuse only parallel loops.

### CODE REVIEW

**Review by rengolin — CHANGES_REQUESTED** [review]
New passes need to have a place in the default pipeline. We need to add an integration test that shows outputs are compatible with the alternative.

**Comment by chelini** [comment]
> New passes need to have a place in the default pipeline. We need to add an integration test that shows outputs are compatible with the alternative.

It is not a new pass. Is the fusion pass with some adjustments to preserve parallelism

**Comment by rengolin** [comment]
And is the fusion pass in the default pipeline?

**Comment by chelini** [comment]
> And is the fusion pass in the default pipeline?

Nope.

**Comment by rengolin** [comment]
> Nope.

Then it should be. And integration tests should demonstrate compatible results with and without parallel loops, via `tpp-run`.

**Comment by chelini** [comment]
> > Nope.
> 
> Then it should be. And integration tests should demonstrate compatible results with and without parallel loops, via `tpp-run`.

Right, I would avoid doing this in the PR. Adding fusion requires a bit of thinking and non-trivial work on the tests. Also keep in mind that at the moment there is not lowering to scf.parallel for forall. I am adding it in MLIR

**Comment by rengolin** [comment]
> Right, I would avoid doing this in the PR. Adding fusion requires a bit of thinking and non-trivial work on the tests. Also keep in mind that at the moment there is not lowering to scf.parallel for scf.forall. I am adding it in MLIR

The problem now is that there is a lot of work going into this pass and when I try to use it it doesn't work unless you fiddle with it a lot. There is no evidence that any of this works in general, even if it does for small, localised examples.

Without this pass running on our existing integration tests, changing what they have to and not changing what they shouldn't, there's no way to know how robust any of this is. The further we work on this pass outside of the default pipeline, the more work we may be throwing it away when we realise things don't fit our existing model.

The work to move the block, tile and fusion pass in the pipeline is more important than further enabling parallelism. Also, @KavithaTipturMadhu is enabling parallelism in a slightly different way, which we need to make sure is compatible with this way. But only running them in the same pipeline, across the same and different examples, that we'll see what we want for what input IR.

**Comment by chelini** [comment]
> > Right, I would avoid doing this in the PR. Adding fusion requires a bit of thinking and non-trivial work on the tests. Also keep in mind that at the moment there is not lowering to scf.parallel for scf.forall. I am adding it in MLIR
> 
> The problem now is that there is a lot of work going into this pass and when I try to use it it doesn't work unless you fiddle with it a lot. There is no evidence that any of this works in general, even if it does for small, localised examples.
> 
> Without this pass running on our existing integration tests, changing what they have to and not changing what they shouldn't, there's no way to know how robust any of this is. The further we work on this pass outside of the default pipeline, the more work we may be throwing it away when we realise things don't fit our existing model.
> 
> The work to move the block, tile and fusion pass in the pipeline is more important than further enabling parallelism. Also, @KavithaTipturMadhu is enabling parallelism in a slightly different way, which we need to make sure is compatible with this way. But only running them in the same pipeline, across the same and different examples, that we'll see what we want for what input IR.

I just talked with @KavithaTipturMadhu and what she is doing is replacing each scf.for to scf.parallel after bufferization. This work with the assumption that all scf.for are actually parallel. I think this PR should be the way forward.

**Comment by rengolin** [comment]
> I just talked with @KavithaTipturMadhu and what she is doing is replacing each scf.for to scf.parallel after bufferization. This work with the assumption that all scf.for are actually parallel. I think this PR should be the way forward.

Do you mean "in addition to" or "completely replacing what she's done"? Either way, we need to discuss this more in depth soon.

**Comment by chelini** [comment]
> > I just talked with @KavithaTipturMadhu and what she is doing is replacing each scf.for to scf.parallel after bufferization. This work with the assumption that all scf.for are actually parallel. I think this PR should be the way forward.
> 
> Do you mean "in addition to" or "completely replacing what she's done"? Either way, we need to discuss this more in depth soon.

completely replacing. I agree.

**Comment by chelini** [comment]
@rengolin I discussed these changes in the morning with @KavithaTipturMadhu and we think would be nice to land in main. Kavitha, can then reduce the diff and some workarounds she did in her mlp branch. I expect this to be the way forward for the default pipeline too.

File: test/Models/mobilenet-without-batchnorm.mlir [hunk 1]
```diff
@@ -1381,14 +1381,14 @@ func.func @mobilenet(%arg0: tensor<1x224x224x3xf32>) -> tensor<1x1001xf32> {
     linalg.yield %1307 : f32
   } -> tensor<1x1x1x1001xf32>
   //
-  // CHECK: %[[subview:.*]] = memref.subview
-  // CHECK: %[[subview1:.*]] = memref.subview
-  // CHECK: %[[subview2:.*]] = memref.subview
+  // C_HECK: %[[subview:.*]] = memref.subview
```
Comment (rengolin) [inline test/Models/mobilenet-without-batchnorm.mlir:1384]: Honestly, I think we can get rid of these CHECKs entirely. Just leave the xsmm call and it should be enough for these entire models.

File: test/Models/resnet50v1.mlir [hunk 2]
```diff
@@ -119,15 +119,14 @@ func.func @resnet50v1(%arg0: tensor<1x224x224x3xf32>) -> tensor<1x1000xf32> {
   %cst_15 = arith.constant dense<0.000000e+00> : tensor<1x56x56x64xf32>
   %cst_16 = arith.constant dense<0.000000e+00> : tensor<1x112x112x64xf32>
 
-  //
+  // TODO: Why are we checking these?????????
```
Comment (rengolin) [inline test/Models/resnet50v1.mlir:122]: To add variable checks to the xsmm calls? I think for these whole models we don't really need that detailed checks. We should save the complexity for the `Convert` and `Integration` checks.

File: include/TPP/Passes.td [hunk 3]
```diff
@@ -207,7 +207,7 @@ def TileConsumerAndFuseProducers : Pass<"tile-consumer-and-fuse-producers",
            "Get producers till maxDepth">,
     Option<"startFromLastFusableConsumer", "start-from-last-consumer", "bool",
            "true", "Fuse from the last fusable consumer of the current target">,
-    Option<"useForAll", "use-for-all", "bool", "false", "Use parallel forAll">
+    Option<"useForAll", "use-for-all", "bool", "true", "Use parallel forAll">
```
Comment (rengolin) [inline include/TPP/Passes.td:210]: And this has no impact if we don't use threads/OMP, right? IE, it lowers to simple loops.

Comment (chelini) [reply include/TPP/Passes.td:210]: At the end of the pipeline you have `scf.parallel`. It has no impact if we use OMP.

File: test/Models/resnet50v1.mlir (hunk 2 above)
Comment (chelini) [reply test/Models/resnet50v1.mlir:122]: Most of them are not used (i.e., c0_i64).

File: test/Models/mobilenet-without-batchnorm.mlir (hunk 1 above)
Comment (chelini) [reply test/Models/mobilenet-without-batchnorm.mlir]: Got rid of these specific CHECKs, but did not touch the rest.

File: lib/TPP/TileConsumerAndFuseProducers.cpp [hunk 4]
```diff
@@ -33,6 +33,89 @@ using namespace mlir;
 
 namespace {
 
+// Convert scf.for to scf.forall after fusion.
+struct ConvertToForAll : public OpRewritePattern<scf::ForOp> {
```
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:37]: We need to rewrite the logic in the fusion pass which is a bit of work. Most of the upstream infrastructure does not work out of the box with forall

File: lib/TPP/TileConsumerAndFuseProducers.cpp [hunk 5]
```diff
@@ -378,6 +464,12 @@ fuseWithEltwise(RewriterBase &rewriter, TilingInterface consumer,
     alreadyFusedOps.insert(tiledOp);
     LLVM_DEBUG(llvm::dbgs() << "NEW OP: " << tiledOp << "\n");
   }
+  if (!tilingResult->loops.empty()) {
```
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:467]: yes

File: test/Models/multi-head-attention.mlir [hunk 6]
```diff
@@ -233,56 +177,37 @@ func.func @multi_head_attention(
     %103 = tensor.empty() : tensor<64x8x8xf32>
     %104 = linalg.fill ins(%cst_1 : f32) outs(%103 : tensor<64x8x8xf32>) -> tensor<64x8x8xf32>
     %105 = linalg.batch_matmul ins(%collapsed_25, %collapsed_26 : tensor<64x8x64xf32>, tensor<64x64x8xf32>) outs(%104 : tensor<64x8x8xf32>) -> tensor<64x8x8xf32>
-    //
```
[the remark this answers is not in the document]
Comment (chelini) [reply test/Models/multi-head-attention.mlir:236]: Yes, linalg.batched_matmul is still there but there is not lowering to tpp.

**Comment by chelini** [comment]
> So, f I get it right, at the mapping phase we block and tile with `scf.for`, then quickly convert to `scf.forall`, and at the conversion phase we change those with `scf.parallel`?
> 
> This looks cleaner than the previous approach, but we need to make sure we don't convert all `scf.forall`, in case we inject our passes into some other framework (OpenXLA) and they also lower `scf.forall` for other purposes.
> 
> But, in the same argument as why we don't want to create a pass that splits `linalg.generic`, we shouldn't be creating a blanket pass that converts `scf.for` into `scf.forall`, when we could be generating `scf.forall` in the first place.
> 
> Not only it's more work that the compiler has to do, and more maintenance that we have to do, but also it could match to patterns from "foreign" IR (when running on another framework) that we shouldn't be touching.

There is no lowering upstream for scf.forall on the CPU side, so we would need to still convert if other framework emit scf.forall.

File: lib/TPP/TileConsumerAndFuseProducers.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:37]: If you want to tile and fuse using scf.forall you need to rewrite most of the pass. `tileAndFuseProducerOfSlice` for example is specific to scf.for. Upstream there are multiple efforts on fusion and is a bit messy at the moment. None as far as I know is working on unify these. Of course fixing upstream would be better but I don't want to entangle myself with it at the moment.

[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:37]: > tileAndFuseProducerOfSlice for example is specific to scf.for

I see, good point.

> Of course fixing upstream would be better but I don't want to entangle myself with it at the moment.

Agreed.

**Review by rengolin — APPROVED** [review]

File: lib/TPP/TileConsumerAndFuseProducers.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/TileConsumerAndFuseProducers.cpp:37]: BTW: I tried to just use scf.forall but I did not get far: https://github.com/chelini/tpp-sandbox/commit/cc19c320a779d455012e6a9e2c20203763efe53d 
I could reuse only the logic to compute the working list. The rest started to diverge really soon. Forall has its own tiling utils (tileToForallOpUsingTileSizes) and `fuseProducerInto` as far as I know is the only way to move producers into the forall. I did not get to the same end-result I was getting with sc.for, as there are some limitations with `fuseProducerInto`

**Comment by rengolin** [comment]
This PR has reduced the performance of our MLP benchmark by ~20%: https://buildkite.com/intel/tpp-mlir/builds/1579#0186e5dc-7c42-468b-a21f-7ca2229cc654

**Pull request merged**